agenthub/codeact_agent/codeact_agent.py

```diff
@@ -177,49 +170,55 @@ def step(self, state: State) -> Action:
         - AgentFinishAction() - end the interaction
         """
+        messages: list[dict[str, str]] = [
+            {'role': 'system', 'content': self.system_message},
+            {
+                'role': 'user',
+                'content': f"Here is an example of how you can interact with the environment for task solving:\n{EXAMPLES}\n\nNOW, LET'S START!",
+            },
+        ]
 
-        updated_info = state.updated_info
-        if updated_info:
-            for prev_action, obs in updated_info:
-                if (
-                    isinstance(prev_action, MessageAction)
-                    and prev_action.source == 'user'
-                ):
-                    self.messages.append(
-                        {'role': 'user', 'content': prev_action.content}
-                    )
-                    if prev_action.content.strip() == '/exit':
-                        # User wants to exit
-                        return AgentFinishAction()
-
-                if isinstance(obs, CmdOutputObservation):
-                    content = 'OBSERVATION:\n' + truncate_observation(obs.content)
-                    content += f'\n[Command {obs.command_id} finished with exit code {obs.exit_code}]]'
-                    self.messages.append({'role': 'user', 'content': content})
-                elif isinstance(obs, IPythonRunCellObservation):
-                    content = 'OBSERVATION:\n' + obs.content
-                    # replace base64 images with a placeholder
-                    splitted = content.split('\n')
-                    for i, line in enumerate(splitted):
-                        if '![image](data:image/png;base64,' in line:
-                            splitted[i] = (
-                                '![image](data:image/png;base64, ...) already displayed to user'
-                            )
-                    content = '\n'.join(splitted)
-                    content = truncate_observation(content)
-                    self.messages.append({'role': 'user', 'content': content})
-                elif isinstance(obs, BrowserOutputObservation):
-                    content = 'OBSERVATION:\n' + truncate_observation(obs.content)
-                    self.messages.append({'role': 'user', 'content': content})
-
-        latest_user_message = [m for m in self.messages if m['role'] == 'user'][-1]
+        for prev_action, obs in state.history:
+            if isinstance(prev_action, MessageAction) and prev_action.source == 'user':
+                if prev_action.content.strip() == '/exit':
+                    return AgentFinishAction()
+                messages.append({'role': 'user', 'content': prev_action.content})
+            else:
+                messages.append(
+                    {
+                        'role': 'user' if prev_action.source == 'user' else 'system',
+                        'content': self.action_to_str(prev_action),
+                    }
+                )
```

Comment:
I think this changes the behavior compared to the previous behavior. The `else` statement won't only handle `MessageAction`s, but also handle other varieties of action where the source is not `user`, right?
Because of this, it seems that `CmdOutputObservation`, `IPythonRunCellObservation`, and `BrowserOutputObservation` be added to messages twice? It seems like they would be added in the else statement here, and also in the statements below.

Maybe I'm misunderstanding something? But either way it'd be nice to clean up this logic to make it more easy to follow.

Reply:
This should be more or less the same as it was before.

Before, when CodeAct took an action, it appended a message to its internal state before returning the action back to the controller. Now, we are rebuilding `messages` on each step, so this `else` is what makes sure CodeAct's previous actions are captured in `messages`.

Good call on cleanup though, this loop was getting a little long! Fixed it up a bit